### PR TITLE
use fmt::join() for printing ranges

### PR DESCRIPTION
**Pull request by tchaikov**
this series intends to deprecate `::join()`, as it always materializes a range into a concrete string. but what we always want is to print the elements in the given range to stream, or to a seastar logger, which is backed by fmtlib. also, because fmtlib offers exactly the same set of features implemented by to_string.hh, this change would allow us to use fmtlib to replace to_string.hh for better maintainability, and potentially better performance. as fmtlib is lazy evaluated, and claims to be performant under most circumstances.

### CODE REVIEW

**Comment by nyh** [comment]
A general request on commit messages: Please try to explain the motivation for every PR, even if you think it's obvious. It usually isn't :-(

In this PR, the only text you wrote in the cover letter was "use fmt::join() for printing ranges". Is this PR about the ability to print ranges, or we were able to print ranges but we didn't use fmt::join() for that? In either case, why is the change interesting? Why do you want to print ranges, or why is fmt::join() better than something else?

**Comment by nyh** [comment]
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4930/

Failed to compile build/debug/cql3/restrictions/statement_restrictions.o with your patch - please check it.

**Comment by tchaikov** [comment]
> A general request on commit messages: Please try to explain the motivation for every PR, even if you think it's obvious. It usually isn't :-(
> 
> In this PR, the only text you wrote in the cover letter was "use fmt::join() for printing ranges". Is this PR about the ability to print ranges, or we were able to print ranges but we didn't use fmt::join() for that? In either case, why is the change interesting? Why do you want to print ranges, or why is fmt::join() better than something else?

@nyh i included the rationales in the commit message but not in the cover letter. if a PR contains multiple commits, am i supposed to copy all of the commit messages into the cover letter, or create a summary in the PR?

File: cql3/restrictions/statement_restrictions.cc
Comment (nyh) [inline cql3/restrictions/statement_restrictions.cc:599]: What does this transformation do, and why is it suddenly needed?
It seems to me that it would print the columns as pointers - but why would we want to do that? We should print the column names, not pointers. What am I missing?

File: gms/versioned_value.hh
Comment (nyh) [inline gms/versioned_value.hh:98]: Is the cast to string_view really necessary here?

File: mutation/atomic_cell.cc
Comment (nyh) [inline mutation/atomic_cell.cc:197]: If the intention is to write directly into os and not even use cell_value_string_builder, you should probably get rid of cell_value_string_builder and not keep it an empty relic.

File: utils/to_string.hh
Comment (nyh) [inline utils/to_string.hh:45]: Why does Scylla need the utils::internal namespace? It's not really a general-purpose library, so everything in it is "internal", isn't it?

**Review by nyh — CHANGES_REQUESTED** [review]

File: gms/versioned_value.hh
Comment (tchaikov) [reply gms/versioned_value.hh:98]: yes. `versioned_value::DELIMITER_STR` is defined as an array of char, i.e., 
```c++
constexpr const char versioned_value::DELIMITER_STR[];
```
which is not a string-like thing. we can improve this part later. but i wanted to focus on the `fmt::join()` related changes in this PR.

File: mutation/atomic_cell.cc
Comment (tchaikov) [reply mutation/atomic_cell.cc:197]: yeah, the general intention is to use `fmt::join()` for better performance and more readable code, and remove `to_string()` helpers in longer run. but i want to make do this in an iterative way -- just focus on one thing if possible.

**Comment by nyh** [comment]
> @nyh i included the rationales in the commit message but not in the cover letter. if a PR contains multiple commits, am i supposed to copy all of the commit messages into the cover letter, or create a summary in the PR?

If the PR has completely unrelated changes, it should be several separate PRs. If, however, the PR has some sort of common theme which is why you chose to make it one PR), you should describe this theme (no need to explain each of the individual patch).

Think of a reviewer (like me) who goes into this PR, having no idea why it even exists. There is an ambigous title, no text and no "Fixes". It have no idea what I'm getting into (although, of course, I can start figuring it out after reading everything, like I did here, but until the last patch I was still in the dark what is the overall aim of this PR and what I can expect to see in the following patches).

In this PR, I think the theme is twofold:
1. You want to move as much as possible from using operator<< to using fmt. I am curious why and would have liked to see an explanation in the commit message - is there a compile-time / run-time / readability benefit? Is this part of a long-term effort to get rid of operator<< in Scylla (do we have an issue?)? Is this patch removing all uses of operator<< from Scylla or just a random 1% in places which also happened to use join()?
2. You want to use fmt::join instead of join() when printing stuff using fmt (now that you use fmt instead of operator<< for more printouts).

File: cql3/restrictions/statement_restrictions.cc
Comment (tchaikov) [reply cql3/restrictions/statement_restrictions.cc:599]: > What does this transformation do, and why is it suddenly needed?
> It seems to me that it would print the columns as pointers - but why would we want to do that? We should print the column names, not pointers. What am I missing?

very good point indeed. because of the exactly the same reason, fmtlib does not format typed pointer, but it format `void*`. and before this change, what gets printed _is_ the the pointers. i don't intend to change the behavior in this change, so added the weird looking transformation. i will add this to the commit message after fixing the FTBFS -- i am not able to reproduce it so far.

Comment (nyh) [reply cql3/restrictions/statement_restrictions.cc:599]: If you believe this is a pre-existing bug (which sounds very suspicious - didn't the person who added this message see it even once?), please at least add a FIXME saying this is a bug that should be fixed. Otherwise it makes this look very deliberate.

Comment (tchaikov) [reply cql3/restrictions/statement_restrictions.cc:599]: i believe. but i am not 100% sure it's a "bug". i'd defer this to @cvybhu .

**Comment by tchaikov** [comment]
> > @nyh i included the rationales in the commit message but not in the cover letter. if a PR contains multiple commits, am i supposed to copy all of the commit messages into the cover letter, or create a summary in the PR?
> 
> If the PR has completely unrelated changes, it should be several separate PRs. If, however, the PR has some sort of common theme which is why you chose to make it one PR), you should describe this theme (no need to explain each of the individual patch).
> 
> Think of a reviewer (like me) who goes into this PR, having no idea why it even exists. There is an ambigous title, no text and no "Fixes". It have no idea what I'm getting into (although, of course, I can start figuring it out after reading everything, like I did here, but until the last patch I was still in the dark what is the overall aim of this PR and what I can expect to see in the following patches).
> 
> In this PR, I think the theme is twofold:
> 
> 1. You want to move as much as possible from using operator<< to using fmt. I am curious why and would have liked to see an explanation in the commit message - is there a compile-time / run-time / readability benefit?

i think you are using commit message and cover letter interchangeability. IMHO, they are different things. 

please refer to the commit message and the cover letter i just revised. and please note, this is but a side effect. despite that personally, i don't like operator<<(), i don't intend to "move as much as possible from using operator<< to using fmt", otherwise it'd be a huge change. i try to be more conservative here.

>  Is this part of a long-term effort to get rid of operator<< in Scylla (do we have an issue?)? Is this patch removing all uses of operator<< from Scylla or just a random 1% in places which also happened to use join()?

if i want to do this, i'd put this explicitly in the commit message. but i didn't. `join()` is the thing i want to kill, as the commit message puts.

> 2. You want to use fmt::join instead of join() when printing stuff using fmt (now that you use fmt instead of operator<< for more printouts).

yes, again. please read the commit message if you haven't done so.

File: cql3/restrictions/statement_restrictions.cc
Comment (cvybhu) [reply cql3/restrictions/statement_restrictions.cc]: :+1: looks like a bug to me.
It has been this way ever since 2015: 73421cb413720c1e8526f7ac1393102e4efe29ac
`test/boost/restrictions_test/token` doesn't check column names, just the message and nobody noticed.

Comment (tchaikov) [reply cql3/restrictions/statement_restrictions.cc]: @cvybhu hi Jan, thanks for ack. added FIXMEs, accordingly.

Comment (cvybhu) [reply cql3/restrictions/statement_restrictions.cc]: Opened #13168

**Comment by tchaikov** [comment]
thank you for your guidance Nadav. i will try to create more helpful cover letters in my future pull requests.

File: mutation/atomic_cell.cc
Comment (nyh) [reply mutation/atomic_cell.cc:197]: So I don't understand, is cell_value_string_builder still used in this code after you stopped using it in those lines you modified here?

**Review by nyh — APPROVED** [review]

File: mutation/atomic_cell.cc
Comment (nyh) [reply mutation/atomic_cell.cc]: Good.
This is exactly what scares me every time I see one of these code-cleanup patches: Among 100 innocent changes there may be one bug hiding, and I don't trust our test suite enough to trust that it will catch any regression. Especially not in the area of error messages (where tests often, rightly, don't insist on the exact text of the error message).

**Comment by tchaikov** [comment]
changelog

- print out the values not the pointers. fix the test failure of
   ```
   -Error from server: code=2200 [Invalid query] message="Columns "ColumnDefinition{name=a, type=org.apache.cassandra.db.marshal.Int32Type, kind=PARTITION_KEY, componentIndex=0, droppedAt=-9223372036854775808}" cannot be restricted by both a normal relation and a token relation"
  +Error from server: code=2200 [Invalid query] message="Columns "0x601004ec5180" cannot be restricted by both a normal relation and a token relation"
  ```
- pass `fmt::join(column_names...)` to `e.execute_cql(...)`, instead of passing `column_names` to it. this should fix the test failure of
  ```
  exceptions::syntax_exception: line 1:16 no viable alternative at input '['
  ```

File: cql3/restrictions/statement_restrictions.cc
Comment (tchaikov) [reply cql3/restrictions/statement_restrictions.cc]: @cvybhu @nyh hi Jan and Nadav, sorry guys. i was wrong. see https://github.com/scylladb/scylladb/commit/9c4c9eff50f0d62d63f0be41b58c0f3044ea234a . i didn't realize that we have this overload, and the hypothesis of "printing a vector of pointers would format the pointer as pointer" is just wrong.

@cvybhu hi Jan, could you please close the #13168. it was a false alarm.

**Review by avikivity — APPROVED** [review]

**Comment by tchaikov** [comment]
the FTBFS on next is due to #12799 which introduced `co_await coroutine::return_exception(std::runtime_error(join("; ", errors)));`, and this PR was merged 9 hours ago.

**Comment by tchaikov** [comment]
changelog

- rebased against master HEAD
- fixed the FTBFS by updating to adapt to change in `service/storage_service.cc` introduced by #12799

**Comment by tchaikov** [comment]
@scylladb/scylla-maint hi scylla maintainer, could you re-enqueue this PR? otherwise the current next would have FTBFS.

**Comment by denesb** [comment]
Dequeued as it breaks the build. Let's wait until promotion runs, then rebase on master.

**Comment by denesb** [comment]
Why would next not build without this PR?

**Comment by denesb** [comment]
> Why would next not build without this PR?

Oh, you mean dequeue and queue again. I prefer waiting until the conflicting patch is merged to master. Rebasing PRs onto next is problematic and is best avoided. If there is a conflict, it is best to wait until next is promoted then rebase onto master.

**Comment by tchaikov** [comment]
the conflicting patch is already in master. anyway, let's wait for CI/CD (a.k.a. promotion) finishes.

**Comment by tchaikov** [comment]
> Dequeued as it breaks the build. Let's wait until promotion runs, then rebase on master.

already rebased.

**Comment by tchaikov** [comment]
rebase again.

**Review by bhalevy — APPROVED** [review]

**Comment by tchaikov** [comment]
thanks Benny. but this failure scares me: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/50/ . i'd like to retry it locally to see if it is reproducible, or at least wait until promoter completes the sanity tests.

**Pull request closed without merging**
